### PR TITLE
Make Funsor compatible with Python 3.5

### PR DESCRIPTION
**Pull request by eb8680**
Necessary for use with Pyro, which supports Python 3.5 and uses it in Travis.

As far as I can tell, this PR contains all of the changes necessary to support Python 3.5 - I couldn't find any lurking order dependence in the obvious places, especially the distribution wrappers.  The only substantive issue I found and fixed was caused by a mysterious nondeterministic edge case in `multipledispatch`, perhaps related to #243.  Otherwise, most of the changes are either removals of f-strings or replacements of dictionaries with `OrderedDict`s in test case lists to make pytest test collection deterministic.

### CODE REVIEW

**Comment by fritzo** [comment]
bye bye f-strings 😢

**Comment by fritzo** [comment]
This is going to be a significant review effort because we have started using the ordered dictionary assumption, but Python 3.5 doesn't follow ordered dictionary semantics. I think we'll need to replace `dict` with `OrderedDict` in many places.

**Comment by eb8680** [comment]
> we have started using the ordered dictionary assumption

I think it's mostly confined to the tests and perhaps to `funsor.distributions`, hopefully we can fix the obvious cases here and deal with others as they come up.

File: setup.py [hunk 1]
```diff
@@ -32,7 +32,7 @@
     },
     author='Uber AI Labs',
     author_email='fritzo@uber.com',
-    python_requires=">=3.6",
+    python_requires=">=3.5",
```
Comment (fritzo) [inline setup.py:35]: Could you also update the `Programming Language :: Python :: 3.6` classifier below?

File: test/test_gaussian.py [hunk 2]
```diff
@@ -207,17 +207,17 @@ def test_smoke(expr, expected_type):
...
-    {'x': reals(2, 3)},
-    {'x': reals(), 'y': reals()},
-    {'x': reals(2), 'y': reals(3)},
-    {'x': reals(4), 'y': reals(2, 3), 'z': reals()},
+    OrderedDict([('x', reals())]),
+    OrderedDict([('x', reals(4))]),
+    OrderedDict([('x', reals(2, 3))]),
+    OrderedDict([('x', reals()), ('y', reals())]),
+    OrderedDict([('x', reals(2)), ('y', reals(3))]),
+    OrderedDict([('x', reals(4)), ('y', reals(2, 3)), ('z', reals())]),
```
Comment (fritzo) [inline test/test_gaussian.py:220]: nit: I think we can revert these changes. I had set this test up so that we could write concise `{}`-dicts in the param statements, then convert them to `OrderedDict` via `OrderedDict(list(sorted(xxx.items())))` below. I think this is still nice to do for readability.

Please let me know if I've misunderstand some issue with pytest collection and the `OrderedDict` change is required.

File: test/test_gaussian.py [hunk 3]
```diff
@@ -231,17 +231,17 @@ def test_align(int_inputs, real_inputs):
 
 
 @pytest.mark.parametrize('int_inputs', [
-    {},
-    {'i': bint(2)},
-    {'i': bint(2), 'j': bint(3)},
+    OrderedDict(),
```
Comment (fritzo) [inline test/test_gaussian.py:234]: ditto here and in all subsequent tests

File: test/test_terms.py [hunk 4]
```diff
@@ -45,6 +45,11 @@
 np.seterr(all='ignore')
 
 
+# have to make this deterministic for pytest collection to work
```
Comment (fritzo) [inline test/test_terms.py:48]: Could you explain the type of error you saw, so I can keep an eye out for similar errors in the future?

File: funsor/registry.py [hunk 5]
```diff
@@ -50,8 +51,11 @@ def register(self, key, *types):
         register = self.registry[key].register
         if self.default:
             objects = (object,) * len(types)
-            if objects != types:
-                register(*objects)(self.default)
+            try:
+                if objects != types and supercedes(types, objects):
+                    register(*objects)(self.default)
+            except TypeError:
+                pass  # mysterious source of ambiguity in Python 3.5 breaks this
```
Comment (eb8680) [inline funsor/registry.py:58]: This is the only significant change in this PR.  It places stricter constraints on method resolution order in multiple dispatch.  If this is reverted, attempting to run the test suite in Python 3.5 will lead to a bunch of strange, nondeterministic and difficult-to-diagnose test failures ultimately caused by patterns not firing when they should.

File: test/test_gaussian.py (hunk 2 above)
Comment (eb8680) [reply test/test_gaussian.py:220]: The `OrderedDict` changes to test parameters in this PR are required for using `pytest-xdist`.  Otherwise, test collection will fail with multiple workers.

File: test/test_terms.py (hunk 4 above)
Comment (eb8680) [reply test/test_terms.py:48]: `pytest-xdist` seems to require all workers to have identical test cases that appear in the same order.  Some test cases in this file are generated by looping over the keys of `REDUCE_OP_TO_NUMERIC`, which is a dictionary.  If you revert this change and attempt to run the tests in Python 3.5 with multiple workers (`pytest -n auto`), test collection will nondeterministically fail because workers may not agree.

File: funsor/sum_product.py [hunk 6]
```diff
@@ -267,7 +266,7 @@ def shift_funsor(f, t):
     if not lags:
         return naive_sequential_sum_product(sum_op, prod_op, trans, time_var, {})
 
-    period = int(np.lcm.reduce(list(lags)))
+    period = int(reduce(lambda a, b: a * b // gcd(a, b), list(lags)))
```
Comment (eb8680) [inline funsor/sum_product.py:269]: `np.lcm.reduce` is not available in the latest version of Numpy that supports Python 3.5.

File: funsor/registry.py (hunk 5 above)
Comment (fritzo) [reply funsor/registry.py:58]: Weird. Do you know whether the Type error occurs in `!=`, `supercedes()`, `and`, or `register()()`?

File: test/test_gaussian.py (hunk 2 above)
Comment (fritzo) [reply test/test_gaussian.py:220]: Thanks for explaining! I'll avoid this pattern in the future.

File: funsor/registry.py (hunk 5 above)
Comment (eb8680) [reply funsor/registry.py:58]: Adding `supercedes` here fixes the test failures, but introduces occasional `TypeError`s which can be disregarded.
